core/src/main/java/tc/oc/pgm/modules/StatsMatchModule.java

```diff
@@ -0,0 +1,299 @@
+                  new PersonalizedText(Integer.toString(deaths), ChatColor.RED),
+                  new PersonalizedText(kd, ChatColor.GREEN))
+              .render());
+    }
+  }
+
+  private static Map<UUID, PlayerStats> allPlayerStats;
+
+  public StatsMatchModule() {
+    allPlayerStats = new HashMap<>();
+  }
+
+  @EventHandler
+  public void onPlayerDeath(MatchPlayerDeathEvent event) {
+    MatchPlayer victim = event.getVictim();
+    MatchPlayer murderer = null;
+    if (event.getKiller() != null) {
+      murderer = event.getKiller().getParty().getPlayer(event.getKiller().getId());
+    }
+    UUID victimUUID = victim.getId();
+    PlayerStats victimStats = allPlayerStats.get(victimUUID);
+
+    if (victimStats == null) {
+      allPlayerStats.put(victimUUID, new PlayerStats());
+      victimStats = allPlayerStats.get(victimUUID);
+    }
+
+    victimStats.onDeath();
+
+    PlayerStats finalVictimStats = victimStats;
+    // Without this delay the "Press LShift to dismount" message from the death minecart will
+    // override the stats message.
+    int victimTaskId =
+        event
+            .getMatch()
+            .getScheduler(MatchScope.LOADED)
+            .runTaskTimer(
+                0,
+                5,
+                () -> {
+                  NMSHacks.sendHotbarMessage(
+                      victim.getBukkit(), finalVictimStats.getBasicStatsMessage());
+                })
+            .getTaskId();
+
+    event
+        .getMatch()
+        .getScheduler(MatchScope.LOADED)
+        .runTaskLater(
+            70,
+            () -> {
+              Bukkit.getScheduler().cancelTask(victimTaskId);
+            });
+
+    if (murderer != null
+        && PlayerRelation.get(victim.getParticipantState(), murderer) != PlayerRelation.ALLY
+        && PlayerRelation.get(victim.getParticipantState(), murderer) != PlayerRelation.SELF) {
+      UUID murdererUUID = murderer.getId();
+      PlayerStats murdererStats = allPlayerStats.get(murdererUUID);
+
+      if (murdererStats == null) {
+        allPlayerStats.put(murdererUUID, new PlayerStats());
+        murdererStats = allPlayerStats.get(murdererUUID);
+      }
+
+      if (event.getDamageInfo() instanceof ProjectileInfo) {
+        murdererStats.setLongestBowKill(
+            victim
+                .getState()
+                .getLocation()
+                .distance(((ProjectileInfo) event.getDamageInfo()).getOrigin()));
+
+        murdererStats.onMurder();
+
+        MatchPlayer finalMurderer = murderer;
+        PlayerStats finalMurdererStats = murdererStats;
+        int murdererTaskId =
+            event
+                .getMatch()
+                .getScheduler(MatchScope.LOADED)
+                .runTaskTimer(
+                    0,
+                    5,
+                    () -> {
+                      NMSHacks.sendHotbarMessage(
+                          finalMurderer.getBukkit(), finalMurdererStats.getBasicStatsMessage());
+                    })
```

Comment:
I’m curious why you chose to call this directly instead of 
`finalMurderer.sendHotbarMessage(finalMurdererStats.getBasicStatsMessage())`?

The [MatchPlayer#sendHotbarMessage](https://github.com/Electroid/PGM/blob/master/core/src/main/java/tc/oc/pgm/api/chat/Audience.java#L61) method calls NMSHacks anyway, so it may be a little more consistent if you like.

Besides that, great job! This will be great once merged 😄

Reply:
Ah! I never knew that, thanks! :)